### PR TITLE
Add authentication

### PR DESCRIPTION
**Pull request by JimmyDore**
## Status
**READY**

## Migrations
YES
a collection 'user' has been added.

## Description
Added an authentication system in the app.
The passwords are encrypted.
Each API call will be redirected to the auth page, if the user is not authenticated.
To check if a user is authenticated, a token is stored in the server cache, and this token is stored in the session (client-side). Each request will send have this token in the request header.
A function has been written to override the XMLHttpRequest calls, and all the calls have been rewritten using this function.
Schemas of the general behaviour are being drawn. 

## Todos
- [x] Tests
- [x] Documentation

## Deploy Notes
Notes regarding deployment the contained body of work.  These should note any
db migrations, etc.

## Steps to Test or Reproduce

Outline the steps to test or reproduce the PR here.

You must have a user to be able to auth a first time.
Add a collection user in the mongodb.
In the router.js file, uncomment the line 441, and comment the line 440 : 
you should be able to add a user via localhost:8080/add-user.html.
(We unlock the route /api/user with POST method) 
Once the user is created, put the lines the way they were before, restart the server, and you can log in to the application.

```sh
git pull --prune
git checkout <feature_branch>
```
## Schemas

general behaviour authentification

authentication

api calls 1

add user

### CODE REVIEW

**Comment by Thomah** [comment]
I had some issues while installing the bcrypt dependency on Windows.
After following [those instructions](https://github.com/kelektiv/node.bcrypt.js/wiki/Installation-Instructions#microsoft-windows), I could execute `npm install` and run the app.

File: public/html/add-user.html [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    <title>Accueil</title>
+    <link rel="icon" href="/favicon.ico" type="image/x-icon" />
+    <link rel="stylesheet" type="text/css" href="boothby.css" />
+  </head>
+
+  <body>
+    <p><a href="index.html"> Home</a> <a href="dialogs.html">Dialogs</a> <a href="config.html">Config</a> <a href="add-user.html">Add user</a></p>
+    <hr />
+    <div>
+        <label for="uname"><b>Username</b></label>
```
Comment (Thomah) [inline public/html/add-user.html]: The value of the for attribute of the label element must be the ID of a non-hidden form control.

File: public/html/add-user.html [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        <input type="text" placeholder="Enter Username" name="uname" id="username" required>
+
+        <label for="psw"><b>Password</b></label>
```
Comment (Thomah) [inline public/html/add-user.html]: The value of the for attribute of the label element must be the ID of a non-hidden form control.

File: public/html/auth.html [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  <head>
+    <meta charset="UTF-8" />
+    <title>Auth</title>
+    <link rel="icon" href="/favicon.ico" type="image/x-icon" />
+    <link rel="stylesheet" type="text/css" href="boothby.css" />
+  </head>
+
+  <body>
+    <div>
+        <label for="uname"><b>Username</b></label>
```
Comment (Thomah) [inline public/html/auth.html]: The value of the for attribute of the label element must be the ID of a non-hidden form control.

File: public/html/auth.html [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        <input type="text" placeholder="Enter Username" name="uname" id="username" required>
+
+        <label for="psw"><b>Password</b></label>
```
Comment (Thomah) [inline public/html/auth.html]: The value of the for attribute of the label element must be the ID of a non-hidden form control.

File: public/js/auth.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
+function send_infos() {
+  var user = document.getElementById("username").value;
+  var pwd = document.getElementById("password").value;
+
+  overload_xhr(
+    "GET",
```
Comment (Thomah) [inline public/js/auth.js]: The method "POST" is preferred for login operation :
[https://stackoverflow.com/questions/5868786/what-method-should-i-use-for-a-login-authentication-request](https://stackoverflow.com/questions/5868786/what-method-should-i-use-for-a-login-authentication-request)

As `/api/user` is already taken, maybe use something like `/api/user/login` for this particular endpoint.

File: src/router.js [hunk 6]
```diff
@@ -59,9 +72,70 @@ var routeStatic = function (request, response) {
...
+        password:request.headers.pwd
+      };
+      api.checkCredentialsUser(credentials, function (data) {
+        if (data != false){ //if the user exists
+          bcrypt.compare(credentials['password'], data['password'], function(err, res) {
+            if (res == true){
+              generated_token = generate_token();
+              username = data['username'];
+              obj = { user: username, token: generated_token };
+             //On stocke dans le serveur en cache le user et le token associé
```
Comment (Thomah) [inline src/router.js]: Nice french comment :smile:

File: public/js/index.js [hunk 7]
```diff
@@ -3,75 +3,78 @@ function reloadIHM() {
...
-    } else {
-      alert(`Request failed.  Returned status of : ${xhr.status}`);
     }
-  };
-  xhr.send();
-}
+  );
+};
 
+//FIX ME : BUG FOUND --> Function not working (even before refactoring) Error: Argument passed in must be a single String of 12 bytes or a string of 24 hex characters
```
Comment (Thomah) [inline public/js/index.js]: The index page is like a playground. Features on this page are not prioritized (maybe in a loooooong time).

File: src/router.js [hunk 8]
```diff
@@ -327,6 +401,32 @@ var waitForChannelsAndIMs = function (callback) {
 };
 
 exports.serve = function (request, response) {
+
+  //If API request, there is a token in the request header which proves that 
+  //   the user is authenticated
+  if (request.headers.token){
+    username = request.headers.username;
```
Comment (Thomah) [inline src/router.js]: Is it possible not to send username in headers ? I think of something like this :
Store every token in a table inside a single spot of the NodeJS cache. Same in the browser cache.

File: src/router.js [hunk 9]
```diff
@@ -327,6 +401,32 @@ var waitForChannelsAndIMs = function (callback) {
...
+    token = request.headers.token;
```
Comment (Thomah) [inline src/router.js:447]: I'm annoying I know. But HTTP Headers are often beginning with an uppercase.. :innocent: 
2019_02_15 - 14_07_01 - accueil

File: src/router.js [hunk 10]
```diff
@@ -59,9 +72,70 @@ var routeStatic = function (request, response) {
...
+        if (data != false){ //if the user exists
+          bcrypt.compare(credentials['password'], data['password'], function(err, res) {
+            if (res == true){
+              generated_token = generate_token();
+              username = data['username'];
+              obj = { user: username, token: generated_token };
+             //On stocke dans le serveur en cache le user et le token associé
+              myCache.set( "Key"+username, obj, function( err, success ){
+                if( !err && success ){
+                  //console.log( success );
```
Comment (Thomah) [inline src/router.js]: A dead code ! Not on my watch !

**Review by Thomah — CHANGES_REQUESTED** [review]
First overview of the PR. It's globally very nice.
Give me a call @JimmyDore if you need to discuss anything :slightly_smiling_face:

**Comment by Thomah** [comment]
Not needed for this PR :
- Creation of a default admin user
- Validation of auth form with the enter key
- Encrypt password on client side

File: public/js/index.js (hunk 7 above)
Comment (JimmyDore) [reply public/js/index.js]: Ok, so I have to remove this comment ?

Comment (Thomah) [reply public/js/index.js]: Well, the whole feature need to be re-thought so... leave your comment as it is and we'll see later :smile:

Comment (JimmyDore) [reply public/js/index.js]: Too laaaate, I add it again

File: public/js/request.js [hunk 11]
```diff
@@ -9,12 +9,10 @@ function overload_xhr(method,
     var xhr = new XMLHttpRequest();
     xhr.open(method, path);
     before_function(xhr);
-    token = sessionStorage.getItem('token');
-    username = sessionStorage.getItem('username');
+    token = sessionStorage.getItem('Token');
```
Comment (Thomah) [inline public/js/request.js]: I didn't pay attention during the first review but the token should be stored in cookies not in session. So when you reopen your browser, the token is preserved as long as the expiration date is not reached. We can have the token in cache expires at the same moment as the browser cookie.
To manage Vanilla JS cookies, you can check this page : https://www.w3schools.com/js/js_cookies.asp

File: public/js/auth.js [hunk 12]
```diff
@@ -3,13 +3,11 @@ function send_infos() {
...
-    "GET",
-    "/api/user",
+    "POST",
+    "/api/user/login",
     function(xhr){
       token = JSON.parse(xhr.response)['token'];
-      username = JSON.parse(xhr.response)['user'];
-      sessionStorage.setItem('username',username);
-      sessionStorage.setItem('token',token);
+      sessionStorage.setItem('Token',token);
```
Comment (Thomah) [inline public/js/auth.js]: Please use cookies instead of session. Checkout the other conversation of the review.

**Review by Thomah — CHANGES_REQUESTED** [review]
We're almost there !
We just need to have the browser token stored in cookies not in session. It should solve our problems of multiple tokens in cache.
I also add to the TODO list (not required for the PR) the feature of disconnection.
As usual, don't hesitate to ping me on Slack if you need anything :slightly_smiling_face:

**Review by Thomah — APPROVED** [review]

**Pull request merged**